Unpickling fails for pattern-bound types

This fails `-Ytest-pickler`:
```scala
object Test {
  def foo(arg: Any) = arg match {
    case bla: List[t] =>
  }
}
```
Because unpickling fails:
```scala
Exception in thread "main" java.lang.AssertionError: assertion failed: Addr(85) Addr(98) Addr(100) BIND
        at scala.Predef$.assert(Predef.scala:170)
        at dotty.tools.dotc.core.tasty.TreeUnpickler$TreeReader.readLengthType$1(TreeUnpickler.scala:268)
        at dotty.tools.dotc.core.tasty.TreeUnpickler$TreeReader.readType(TreeUnpickler.scala:330)
        at dotty.tools.dotc.core.tasty.TreeUnpickler$TreeReader$$anonfun$5.apply(TreeUnpickler.scala:233)
        at dotty.tools.dotc.core.tasty.TreeUnpickler$TreeReader$$anonfun$5.apply(TreeUnpickler.scala:233)
```

Indeed, we pickle three things:
```scala
      else if (sym is Flags.BindDefinedType) {
        registerDef(sym)
        writeByte(BIND)
        withLength {
          pickleName(sym.name)
          pickleType(sym.info)
          pickleRef()
        }
      }
```
Thus getting the tree:
```scala
    85:                 BIND(13) 26[t]
    88:                   TYPEBOUNDS(8)
    90:                     TYPEREF 27[Nothing]
    92:                       SHARED 52
    94:                     TYPEREF 19[Any]
    96:                       SHARED 52
    98:                   TYPEREFdirect 85
```

But we unpickle two things:
```scala
            case BIND =>
              val sym = ctx.newSymbol(ctx.owner, readName().toTypeName, BindDefinedType, readType())
              registerSym(start, sym)
              TypeRef.withFixedSym(NoPrefix, sym.name, sym)
```

--- a/compiler/src/dotty/tools/dotc/core/Mode.scala
+++ b/compiler/src/dotty/tools/dotc/core/Mode.scala
@@ -81,10 +81,13 @@ object Mode {
   /** Read original positions when unpickling from TASTY */
   val ReadPositions = newMode(16, "ReadPositions")
 
+  /** Don't suppress exceptions thrown during show */
+  val PrintShowExceptions = newMode(17, "PrintShowExceptions")
+
   val PatternOrType = Pattern | Type
 
   /** We are elaborating the fully qualified name of a package clause.
    *  In this case, identifiers should never be imported.
    */
-  val InPackageClauseName = newMode(17, "InPackageClauseName")
+  val InPackageClauseName = newMode(18, "InPackageClauseName")
 }

--- a/compiler/src/dotty/tools/dotc/core/tasty/TastyFormat.scala
+++ b/compiler/src/dotty/tools/dotc/core/tasty/TastyFormat.scala
@@ -431,7 +431,8 @@ object TastyFormat {
        | ANNOTATEDtpt
        | ANDtpt
        | ORtpt
-       | BYNAMEtpt => true
+       | BYNAMEtpt
+       | BIND => true
     case _ => false
   }
 

--- a/compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala
+++ b/compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala
@@ -249,6 +249,7 @@ class TreeUnpickler(reader: TastyReader, nameAtRef: NameRef => TermName, posUnpi
             case BIND =>
               val sym = ctx.newSymbol(ctx.owner, readName().toTypeName, BindDefinedType, readType())
               registerSym(start, sym)
+              if (currentAddr != end) readType()
               TypeRef.withFixedSym(NoPrefix, sym.name, sym)
             case POLYtype =>
               readMethodic(PolyType, _.toTypeName)

--- a/compiler/src/dotty/tools/dotc/printing/Formatting.scala
+++ b/compiler/src/dotty/tools/dotc/printing/Formatting.scala
@@ -31,7 +31,8 @@ object Formatting {
       case arg: Showable =>
         try arg.show
         catch {
-          case NonFatal(ex) => s"[cannot display due to $ex, raw string = $toString]"
+          case NonFatal(ex) if !ctx.mode.is(Mode.PrintShowExceptions) =>
+            s"[cannot display due to $ex, raw string = $toString]"
         }
       case _ => arg.toString
     }

--- a/compiler/src/dotty/tools/dotc/transform/Pickler.scala
+++ b/compiler/src/dotty/tools/dotc/transform/Pickler.scala
@@ -79,7 +79,8 @@ class Pickler extends Phase {
           ctx.fresh
             .setPeriod(Period(ctx.runId + 1, FirstPhaseId))
             .setReporter(new ThrowingReporter(ctx.reporter))
-            .addMode(Mode.ReadPositions))
+            .addMode(Mode.ReadPositions)
+            .addMode(Mode.PrintShowExceptions))
     result
   }
 

--- a/tests/pickling/i2437a.scala
+++ b/tests/pickling/i2437a.scala
@@ -1,0 +1,5 @@
+object Test {
+  def foo(arg: Any) = arg match {
+    case bla: List[t] =>
+  }
+}

--- a/tests/pickling/i2437b.scala
+++ b/tests/pickling/i2437b.scala
@@ -1,0 +1,6 @@
+class Foo[T]
+object Test {
+  def foo[T](arg: Foo[T]) = arg match {
+    case bla: Foo[_] =>
+  }
+}

